### PR TITLE
feat: SCSI controller and unit number support, attach disks on build

### LINKED ISSUE #538: diskAttachmentProperties for disk attachment to a VM not exposed
### Code of Conduct

This project has a Code of Conduct that all participants are expected to understand and follow:
* [x] I have read and agree to the project's [Code of Conduct](https://github.com/vmware/terraform-provider-vra/blob/main/CODE-OF-CONDUCT.md)

### Description

The API for attaching a disk to a VM includes diskAttachmentProperties in the schema.  This allows the scsiController and unitNumber for the disk to be set.
These parameters cannot be set via the VRA Provder.

Without these options, block devices are attached to a VM in a random order.  Operations such as resizing and removal via Terraform and ordered mount of the disks inside the VM aren't possible as it is not possible to determine which disk is which.

### Describe alternatives you've considered

NA

### References

https://api.mgmt.cloud.vmware.com/iaas-api/swagger/webjars/swagger-ui/index.html#/Compute/attachMachineDisk

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by ZsoltFejes**
- Feature added based on request: https://github.com/vmware/terraform-provider-vra/issues/538 To be able to specify scsi controller and unit for block device attach
- Added support for attaching disks on build

Closes #538

### CODE REVIEW

**Comment by ZsoltFejes** [comment]
## Tests I carried out
- One disk with valid controller and unit, and attach before boot set to true - Disk is attached during build correctly to the specified controller and unit
- One disk with correct controller and unit, and attach before boot set to true - Disk was attached to the correct controller and unit after the VM was built
- One disk with correct controller but no unit - Default Aria automation behaviour, attached disk to the root disk’s controller and the next available unit
- One disk with missing controller specified but unit is specified - Default aria behaviour, attached disk to the root disk’s controller and the next available unit
- Two disks with the same controller and unit - Aria failed to attach the second disk, returned error stating unable to attach controller and unit number is already attached
- Two disks, one disk was specified with controller and unit number the second disk wasn’t - Disk with specified placement was correct, the one without placement was attached to the controller of the root disk to the next available unit

File: vra/resource_machine.go [hunk 1]
```diff
@@ -92,6 +97,16 @@ func resourceMachine() *schema.Resource {
 							Required:    true,
 							Description: "The id of the existing block device.",
 						},
+						"scsi_controller": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "The id of the scsi controller. Example: SCSI_Controller_0",
```
Comment (tenthirtyam) [inline vra/resource_machine.go:103]:
```suggestion
							Description: "The id of the SCSI controller. Example: SCSI_Controller_0",
```

File: vra/resource_machine.go [hunk 2]
```diff
@@ -92,6 +97,16 @@ func resourceMachine() *schema.Resource {
...
+						},
+						"unit_number": {
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "The unit number of the scsi controller. Example: 2",
```
Comment (tenthirtyam) [inline vra/resource_machine.go:108]:
```suggestion
							Description: "The unit number of the SCSI controller. Example: 2",
```

File: vra/resource_machine.go [hunk 3]
```diff
@@ -116,6 +131,16 @@ func resourceMachine() *schema.Resource {
...
```
Comment (tenthirtyam) [inline vra/resource_machine.go:142]:
```suggestion
							Description: "The unit number of the SCSI controller. Example: 2",
```

File: website/docs/r/vra_machine.html.markdown [hunk 4]
```diff
@@ -130,6 +130,8 @@ Example: `[{"mandatory" : "true", "expression": "environment":"prod"}, {"mandato
 
 * `created_at` - Date when the entity was created. Date and time format is ISO 8601 and UTC.
 
+* `attach_disks_before_boot` - By default, disks are attached after the VM has been built as FCDs cannot be attached to machine as day 0
```
Comment (tenthirtyam) [inline website/docs/r/vra_machine.html.markdown:133]:
```suggestion
* `attach_disks_before_boot` - By default, disks are attached after the VM has been built as FCDs cannot be attached to machine as day 0.
```

File: website/docs/r/vra_machine.html.markdown [hunk 5]
```diff
@@ -138,6 +140,10 @@ Example: `[{"mandatory" : "true", "expression": "environment":"prod"}, {"mandato
 
     * `name` - Human-friendly block-device name used as an identifier in APIs that support this option.
 
+    * `scsi_controller` - The id of the scsi controller. Example: SCSI_Controller_0
+
+    * `unit_number` - The unit number of the scsi controller. Example: 2
```
Comment (tenthirtyam) [inline website/docs/r/vra_machine.html.markdown:145]:
```suggestion
    * `unit_number` - The unit number of the SCSI controller (_e.g_., `2.)
```

File: vra/resource_machine.go [hunk 6]
```diff
@@ -71,6 +71,11 @@ func resourceMachine() *schema.Resource {
 				Computed:    true,
 				Description: "The id of the deployment that is associated with this resource.",
 			},
+			"attach_disks_before_boot": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "By default, disks are attached after the VM has been built as FCDs cannot be attached to machine as day 0",
```
Comment (tenthirtyam) [inline vra/resource_machine.go:77]:
```suggestion
				Description: "By default, disks are attached after the machine has been built. FCDs cannot be attached to machine as a day 0 task.",
```

**Comment by frodenas** [comment]
The change looks good to me. @ZsoltFejes can you please fix the lint errors? Thanks!

**Comment by ZsoltFejes** [comment]
@frodenas Thank you for having a look at this pull request. I have pushed the change with the fix.

**Review by frodenas — APPROVED** [review]

**Pull request merged**